### PR TITLE
Use snapshot information to build searchable snapshot store MetadataSnapshot

### PR DESCRIPTION
**Pull request by tlrx**
While investigating possible optimizations to speed up searchable snapshots shard restores, we noticed that Elasticsearch builds the list of shard files on local disk in order to compare it with the list of files contained in the snapshot to restore. This list of files is materialized with a `MetadataSnapshot` object whose construction involves to read the footer checksum of every files of the shard using `Store.checksumFromLuceneFile()` method.

Further investigation shows that a `MetadataSnapshot` object is also created for other types of operations like building the list of files to recover in a peer recovery (and primary shard relocation)
or in order to assign a shard to a node. These operations use the `Store.getMetadata(IndexCommit)` method to build the list of files and checksums.

In the case of searchable snapshots building the `MetadataSnapshot` object can potentially trigger cache misses, which in turn can cause the download and the writing in cache of the last range of the file in order to check the 16 bytes footer. This in turn can cause more evictions.

Since searchable snapshots already contains the footer information of every file in `BlobStoreIndexShardSnapshot` it can directly read the checksum from it and avoid to use the cache at all to create a `MetadataSnapshot` for the operations mentioned above.

This pull request adds a shortcut to the `SearchableSnapshotDirectory.openInput()` method - similarly to what already exists for segment infos - so that it creates a specific `IndexInput` for checksum reading operation.

### CODE REVIEW

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/checksum/ChecksumBlobContainerIndexInput.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    private int position;
+
+    private ChecksumBlobContainerIndexInput(String name, long length, byte[] checksum, IOContext context) {
+        super("ChecksumBlobContainerIndexInput(" + name + ')');
+        ensureReadOnceChecksumContext(context);
+        assert checksum.length == CodecUtil.footerLength();
+        this.checksum = Objects.requireNonNull(checksum);
+        this.length = length;
+        this.offset = length - this.checksum.length;
+        assert offset > 0;
```
Comment (DaveCTurner) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/checksum/ChecksumBlobContainerIndexInput.java:39]: Should we set `position = -offset` here too? I think it's strange, although probably irrelevant, that a newly-opened index input has a nonzero file pointer.

**Review by DaveCTurner — COMMENTED** [review]
Very neat, I like it. I left a few tiny comments.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/checksum/ChecksumBlobContainerIndexInput.java (hunk 1 above)
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/checksum/ChecksumBlobContainerIndexInput.java:39]: (that would imply that we'd need to throw appropriate exceptions if `position < 0` in `readByte()` and `readBytes()` too)

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/checksum/ChecksumBlobContainerIndexInput.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    @Override
+    public void seek(long pos) throws IOException {
+        if (pos < 0) {
+            throw new IllegalArgumentException("Seeking to negative position: " + pos);
+        } else if (pos > length()) {
+            throw new EOFException("seek past EOF");
+        } else if (pos < offset) {
+            throw new EOFException("Can't read before footer checksum");
```
[the remark this answers is not in the document]
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/checksum/ChecksumBlobContainerIndexInput.java]: Right, that's a bad copy paste.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/checksum/ChecksumBlobContainerIndexInput.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        }
+        position = Math.toIntExact(pos - offset);
+    }
+
+    @Override
+    public IndexInput slice(String sliceDescription, long offset, long length) {
+        throw new UnsupportedOperationException();
```
[the remark this answers is not in the document]
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/checksum/ChecksumBlobContainerIndexInput.java:83]: Ok

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/checksum/ChecksumBlobContainerIndexInput.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    @Override
+    public IndexInput clone() {
+        throw new UnsupportedOperationException();
```
[the remark this answers is not in the document]
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/checksum/ChecksumBlobContainerIndexInput.java:89]: ok

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/checksum/ChecksumBlobContainerIndexInput.java (hunk 1 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/checksum/ChecksumBlobContainerIndexInput.java]: Actually I think it's easier to reason about using a global `position` and adjust it only when reading the checksum bytes. I pushed 709e64424108b0863a96884cc682ad7bbae61d5a please let me know what you think

**Comment by tlrx** [comment]
Thanks @DaveCTurner, I've applied your feedback.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by tlrx** [comment]
Thanks @DaveCTurner !

**Pull request merged**